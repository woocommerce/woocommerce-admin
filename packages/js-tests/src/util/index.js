/**
 * External dependencies
 */
import { screen } from '@testing-library/react';

/**
 * For use with react-testing-library, like getText but allows the text to reside in multiple elements
 *
 * @param {Object} query - Original query.
 *
 * @return {Array} - Array of two arrays, first including truthy values, and second including falsy.
 */
const withMarkup = ( query ) => ( text ) =>
	query( ( content, node ) => {
		const hasText = ( domNode ) => domNode.textContent === text;
		const childrenDontHaveText = Array.from( node.children ).every(
			( child ) => ! hasText( child )
		);

		return hasText( node ) && childrenDontHaveText;
	} );

export const getByTextWithMarkup = withMarkup( screen.getByText );
