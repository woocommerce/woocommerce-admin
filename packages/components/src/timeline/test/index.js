/* eslint-disable jest/no-mocks-import */
/**
 * External dependencies
 */
import { render } from '@testing-library/react';
import { createElement } from '@wordpress/element';

/**
 * Internal dependencies
 */
import Timeline from '..';
import mockData from '../__mocks__/timeline-mock-data';
import { groupItemsUsing, sortByDateUsing } from '../util.js';

describe( 'Timeline', () => {
	test( 'Empty snapshot', () => {
		const { container } = render( <Timeline /> );
		expect( container ).toMatchSnapshot();
	} );

	test( 'With data snapshot', () => {
		const { container } = render( <Timeline items={ mockData } /> );
		expect( container ).toMatchSnapshot();
	} );

	describe( 'Timeline utilities', () => {
		test( 'Sorts correctly', () => {
			const jan21 = new Date( 2020, 0, 21 );
			const jan22 = new Date( 2020, 0, 22 );
			const jan23 = new Date( 2020, 0, 23 );

			const data = [
				{ id: 0, date: jan22 },
				{ id: 1, date: jan21 },
				{ id: 2, date: jan23 },
			];
			const expectedAsc = [
				{ id: 1, date: jan21 },
				{ id: 0, date: jan22 },
				{ id: 2, date: jan23 },
			];
			const expectedDesc = [
				{ id: 2, date: jan23 },
				{ id: 0, date: jan22 },
				{ id: 1, date: jan21 },
			];

			expect( data.sort( sortByDateUsing( 'asc' ) ) ).toStrictEqual(
				expectedAsc
			);
			expect( data.sort( sortByDateUsing( 'desc' ) ) ).toStrictEqual(
				expectedDesc
			);
		} );

		test( "Empty item list doesn't break sort", () => {
			expect( [].sort( sortByDateUsing( 'asc' ) ) ).toStrictEqual( [] );
		} );

		test( "Single item doesn't change on sort", () => {
			const items = [ { date: new Date( 2020, 0, 1 ) } ];
			expect( items.sort( sortByDateUsing( 'asc' ) ) ).toBe( items );
		} );

		test( 'Groups correctly', () => {
			const jan22 = new Date( 2020, 0, 22 );
			const jan23 = new Date( 2020, 0, 23 );
			const items = [
				{ id: 0, date: jan22 },
				{ id: 1, date: jan23 },
				{ id: 2, date: jan22 },
			];
			const expected = [
				{
					date: jan22,
					items: [
						{ id: 0, date: jan22 },
						{ id: 2, date: jan22 },
					],
				},
				{
					date: jan23,
					items: [ { id: 1, date: jan23 } ],
				},
			];

			expect(
				items.reduce( groupItemsUsing( 'days' ), [] )
			).toStrictEqual( expected );
		} );

		test( "Empty item list doesn't break grouping", () => {
			expect( [].reduce( groupItemsUsing( 'days' ), [] ) ).toStrictEqual(
				[]
			);
		} );

		test( 'Single item grouped correctly', () => {
			const jan22 = new Date( 2020, 0, 22 );
			const items = [ { id: 0, date: jan22 } ];
			const expected = [
				{
					date: jan22,
					items: [ { id: 0, date: jan22 } ],
				},
			];
			expect(
				items.reduce( groupItemsUsing( 'days' ), [] )
			).toStrictEqual( expected );
		} );
	} );
} );
