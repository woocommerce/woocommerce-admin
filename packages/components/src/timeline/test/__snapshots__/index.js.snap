// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline Empty snapshot 1`] = `
<div>
  <div
    class="woocommerce-timeline"
  >
    <p
      class="timeline_no_events"
    >
      No data to display
    </p>
  </div>
</div>
`;

exports[`Timeline With data snapshot 1`] = `
<div>
  <div
    class="woocommerce-timeline"
  >
    <ul>
      <li
        class="woocommerce-timeline-group"
      >
        <p
          class="woocommerce-timeline-group__title"
        >
          January 22, 2020
        </p>
        <ul>
          <li
            class="woocommerce-timeline-item"
          >
            <div
              class="woocommerce-timeline-item__top-border"
            />
            <div
              class="woocommerce-timeline-item__title"
            >
              <div
                class="woocommerce-timeline-item__headline"
              >
                <svg
                  class="gridicon gridicons-cross is-error"
                  height="16"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g>
                    <path
                      d="M18.36 19.78L12 13.41l-6.36 6.37-1.42-1.42L10.59 12 4.22 5.64l1.42-1.42L12 10.59l6.36-6.36 1.41 1.41L13.41 12l6.36 6.36z"
                    />
                  </g>
                </svg>
                <span>
                  string in headline
                </span>
              </div>
              <span
                class="woocommerce-timeline-item__timestamp"
              >
                3:13pm
              </span>
            </div>
            <div
              class="woocommerce-timeline-item__body"
            >
              <span>
                <span>
                  span in body
                </span>
              </span>
            </div>
          </li>
        </ul>
        <hr />
      </li>
      <li
        class="woocommerce-timeline-group"
      >
        <p
          class="woocommerce-timeline-group__title"
        >
          January 20, 2020
        </p>
        <ul>
          <li
            class="woocommerce-timeline-item"
          >
            <div
              class="woocommerce-timeline-item__top-border"
            />
            <div
              class="woocommerce-timeline-item__title"
            >
              <div
                class="woocommerce-timeline-item__headline"
              >
                <svg
                  class="gridicon gridicons-refresh is-warning"
                  height="16"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g>
                    <path
                      d="M17.91 14c-.478 2.833-2.943 5-5.91 5-3.308 0-6-2.692-6-6s2.692-6 6-6h2.172l-2.086 2.086L13.5 10.5 18 6l-4.5-4.5-1.414 1.414L14.172 5H12a8 8 0 000 16c4.079 0 7.438-3.055 7.931-7H17.91z"
                    />
                  </g>
                </svg>
                <span>
                  <span>
                    span in headline
                  </span>
                </span>
              </div>
              <span
                class="woocommerce-timeline-item__timestamp"
              >
                11:45pm
              </span>
            </div>
            <div
              class="woocommerce-timeline-item__body"
            />
          </li>
          <li
            class="woocommerce-timeline-item"
          >
            <div
              class="woocommerce-timeline-item__top-border"
            />
            <div
              class="woocommerce-timeline-item__title"
            >
              <div
                class="woocommerce-timeline-item__headline"
              >
                <svg
                  class="gridicon gridicons-checkmark is-success"
                  height="16"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g>
                    <path
                      d="M9 19.414l-6.707-6.707 1.414-1.414L9 16.586 20.293 5.293l1.414 1.414z"
                    />
                  </g>
                </svg>
                <span>
                  <p>
                    p tag in headline
                  </p>
                </span>
              </div>
              <span
                class="woocommerce-timeline-item__timestamp"
              />
            </div>
            <div
              class="woocommerce-timeline-item__body"
            >
              <span>
                <p>
                  p element in body
                </p>
              </span>
              <span>
                string in body
              </span>
            </div>
          </li>
        </ul>
        <hr />
      </li>
      <li
        class="woocommerce-timeline-group"
      >
        <p
          class="woocommerce-timeline-group__title"
        >
          January 17, 2020
        </p>
        <ul>
          <li
            class="woocommerce-timeline-item"
          >
            <div
              class="woocommerce-timeline-item__top-border"
            />
            <div
              class="woocommerce-timeline-item__title"
            >
              <div
                class="woocommerce-timeline-item__headline"
              >
                <svg
                  class="gridicon gridicons-cross"
                  height="16"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g>
                    <path
                      d="M18.36 19.78L12 13.41l-6.36 6.37-1.42-1.42L10.59 12 4.22 5.64l1.42-1.42L12 10.59l6.36-6.36 1.41 1.41L13.41 12l6.36 6.36z"
                    />
                  </g>
                </svg>
                <span>
                  undefined body and string headline
                </span>
              </div>
              <span
                class="woocommerce-timeline-item__timestamp"
              >
                1:45am
              </span>
            </div>
            <div
              class="woocommerce-timeline-item__body"
            />
          </li>
        </ul>
        <hr />
      </li>
    </ul>
  </div>
</div>
`;
